pkg/datastore/remotedatastore.go

```diff
@@ -0,0 +1,112 @@
+package datastore
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/lthibault/log"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/blocknative/dreamboat/pkg/structs"
+)
+
+type RemoteDatastore interface {
+	GetPayload(context.Context, structs.PayloadKey) (*structs.BlockAndTrace, error)
+	PutPayload(context.Context, structs.PayloadKey, *structs.BlockAndTrace, time.Duration) error
+}
+
+type LocalRemoteDatastore struct {
+	*Datastore
+	Remote RemoteDatastore
+
+	Logger log.Logger
+
+	m LRDatastoreMetrics
+}
+
+func NewLocalRemoteDatastore(local *Datastore, remote RemoteDatastore, l log.Logger) *LocalRemoteDatastore {
+	s := LocalRemoteDatastore{
+		Datastore: local,
+		Remote:    remote,
+		Logger:    l.WithField("relay-service", "stream-datastore"),
+	}
+
+	s.initMetrics()
+
+	return &s
+}
+
+type getPayloadResponse struct {
+	block     *structs.BlockAndTrace
+	isLocal   bool
+	fromCache bool
+	err       error
+}
+
+func (s *LocalRemoteDatastore) GetPayload(ctx context.Context, key structs.PayloadKey) (*structs.BlockAndTrace, bool, error) {
+	timer0 := prometheus.NewTimer(s.m.Timing.WithLabelValues("getPayload", "all"))
+	defer timer0.ObserveDuration()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	responses := make(chan getPayloadResponse, 2)
+
+	go func(ctx context.Context, resp chan getPayloadResponse) {
+		timer1 := prometheus.NewTimer(s.m.Timing.WithLabelValues("getPayload", "local"))
+		block, fromCache, err := s.Datastore.GetPayload(ctx, key)
+		timer1.ObserveDuration()
+		responses <- getPayloadResponse{block: block, isLocal: true, fromCache: fromCache, err: err}
+	}(ctx, responses)
+
+	go func(ctx context.Context, resp chan getPayloadResponse) {
+		timer1 := prometheus.NewTimer(s.m.Timing.WithLabelValues("getPayload", "remote"))
+		block, err := s.Remote.GetPayload(ctx, key)
+		timer1.ObserveDuration()
+		responses <- getPayloadResponse{block: block, isLocal: false, fromCache: false, err: err}
+	}(ctx, responses)
+
+	for i := 0; i < cap(responses); i++ {
+		var resp getPayloadResponse
+		select {
+		case resp = <-responses:
+		case <-ctx.Done():
+			return &structs.BlockAndTrace{}, false, ctx.Err()
```

Comment:
That and other returns further down may just leave us with the channel with unread responses. In this particular case even unclosed :)